### PR TITLE
Added test cases for cartesian_representation and cartesian_differential

### LINKED ISSUE #71: Add new tests for cartesian_representation
Adding tests for cartesian_representation

### LINKED ISSUE #73: Add new tests for cartesian_differential
Adding tests for `cartesian_differential`

### PR DESCRIPTION
**Pull request by sarthak2007**
### Description

Testcases for `cartesian_representation.hpp` and `cartesian_differential.hpp`. Closes #71 and closes #73 

### Tasklist

- [x] Add test case(s)
- [ ] Ensure all CI builds pass
- [ ] Review and approve

### CODE REVIEW

File: test/coordinate/cartesian_differential.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    BOOST_CHECK_CLOSE(motion1.get_dx().value(), 1.5, 0.001);
+    BOOST_CHECK_CLOSE(motion1.get_dy().value(), 9.0, 0.001);
+    BOOST_CHECK_CLOSE(motion1.get_dz().value(), 3, 0.001);
+    
+    //copy constructor
+    auto motion2 = make_cartesian_differential(motion1);
+    BOOST_CHECK_CLOSE(motion1.get_dx().value(), motion2.get_dx().value(), 0.001);
+    BOOST_CHECK_CLOSE(motion1.get_dy().value(), motion2.get_dy().value(), 0.001);
+    BOOST_CHECK_CLOSE(motion1.get_dz().value(), motion2.get_dz().value(), 0.001);
+
```
Comment (Ruchika2696) [inline test/coordinate/cartesian_differential.cpp]: Can you please add tests for checking if the quantity stored in each coordinate is as expected.
you can use [`BOOST_TEST`](https://www.boost.org/doc/libs/1_70_0/libs/test/doc/html/boost_test/testing_tools/boost_test_universal_macro.html)
you can do something like this:
```
BOOST_TEST(std::is_same<decltype(motion1.get_dx()), bu::quantity<bu::divide_typeof_helper<bu::si::length, bu::si::time>::type>>::value)
```

There can be many ways of doing this maybe even more efficient too it's up to you. Add such tests wherever you feel is needed in differential and representation tests especially when any kind of conversions are performed.

**Review by Ruchika2696 — CHANGES_REQUESTED** [review]

**Review by lpranam — CHANGES_REQUESTED** [review]
Can you add tests for default constructors too? you can combine them with `set_x` function and in addition tests including unit checks would be great.

**Review by lpranam — CHANGES_REQUESTED** [review]
Can you also divide constructor tests in smaller test cases? It would be easier to maintain in future.

**Review by lpranam — CHANGES_REQUESTED** [review]
can you also add failing test which indicates invalid conversions?

File: test/coordinate/cartesian_differential.cpp [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    //checking construction from value
+    auto motion1 = make_cartesian_differential
+    (1.5*meters/seconds, 9.0*si::kilo*meters/seconds, 3.0*meters/seconds);
+    BOOST_CHECK_CLOSE(motion1.get_dx().value(), 1.5, 0.001);
+    BOOST_CHECK_CLOSE(motion1.get_dy().value(), 9.0, 0.001);
+    BOOST_CHECK_CLOSE(motion1.get_dz().value(), 3, 0.001);
+
+    //checking whether quantity stored is as expected or not
+    BOOST_TEST((std::is_same<decltype(motion1.get_dx()),
+        quantity<bu::divide_typeof_helper<si::length, si::time>::type>>::value));
```
Comment (Ruchika2696) [inline test/coordinate/cartesian_differential.cpp]: You may replace these with `quantity<si::velocity>`

File: test/coordinate/cartesian_differential.cpp [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    BOOST_TEST((std::is_same<decltype(motion4.get_dy()),
+        quantity<bu::divide_typeof_helper<decltype(si::kilo*meters), si::time>::type>>::value));
+    BOOST_TEST((std::is_same<decltype(motion4.get_dz()),
+        quantity<bu::divide_typeof_helper<si::length, si::time>::type>>::value));
+        
+    // //Conversion from one unit type to other
+    auto motion5 = make_cartesian_differential
+    <double, quantity<si::velocity>, quantity<si::velocity>, quantity<si::velocity>>(motion1);
+    BOOST_CHECK_CLOSE(motion4.get_dx().value(), 1.5, 0.001);
```
Comment (Ruchika2696) [inline test/coordinate/cartesian_differential.cpp]: `motion4` needs to be replaced with `motion5`

File: test/coordinate/cartesian_differential.cpp [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    BOOST_CHECK_CLOSE(motion4.get_dy().value(), 9.0, 0.001);
```
Comment (Ruchika2696) [inline test/coordinate/cartesian_differential.cpp]: `9.0` will change to `9000.0` if converted from km/s to m/s

File: test/coordinate/cartesian_differential.cpp [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+BOOST_AUTO_TEST_SUITE_END()
+
+BOOST_AUTO_TEST_SUITE(cartesian_differential_operators)
+
+BOOST_AUTO_TEST_CASE(addition)
+{
+    auto motion1 = make_cartesian_differential
+        (11.0*meters/seconds, 15.0*meters/seconds, 19.0*meters/seconds);
+    auto motion2 = make_cartesian_differential
```
Comment (Ruchika2696) [inline test/coordinate/cartesian_differential.cpp:218]: this is obvious can you try adding with different units

File: test/coordinate/cartesian_differential.cpp [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+
+    auto product = make_cartesian_differential
+        (motion1 * quantity<si::time>(5*seconds));
+
+    BOOST_CHECK_CLOSE(product.get_dx().value(), 15.0, 0.001);
+    BOOST_CHECK_CLOSE(product.get_dy().value(), 45.0, 0.001);
+    BOOST_CHECK_CLOSE(product.get_dz().value(), 30.0, 0.001);
+
+    //checking whether quantity stored is as expected or not
+    BOOST_TEST((std::is_same<decltype(product.get_dx()),
```
Comment (Ruchika2696) [inline test/coordinate/cartesian_differential.cpp]: if it is velocity is multiplied with time then isn't it supposed to be length only?

File: test/coordinate/cartesian_representation.cpp [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    BOOST_CHECK_CLOSE(point1.get_z().value(), point3.get_z().value(), 0.001);
+
+    //checking whether quantity stored is as expected or not
+    BOOST_TEST((std::is_same<decltype(point3.get_x()), quantity<si::length>>::value));
+    BOOST_TEST((std::is_same<decltype(point3.get_y()), quantity<decltype(si::kilo*meter)>>::value));
+    BOOST_TEST((std::is_same<decltype(point3.get_z()), quantity<decltype(si::centi*meter)>>::value));
+
+    cartesian_representation<double, quantity<si::length>, quantity<si::length>,
+    quantity<si::length>> point4(point1);
+    BOOST_CHECK_CLOSE(point1.get_x().value(), point4.get_x().value(), 0.001);
```
Comment (Ruchika2696) [inline test/coordinate/cartesian_representation.cpp]: there are some implicit conversions happening from km to m and cm to m. How values from both points can be the same?

File: test/coordinate/cartesian_representation.cpp [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+    auto point2 = make_cartesian_representation
+        (3.0 * si::milli * meters, 5.0 * si::centi * meters, 4.0 * meters);
+    //spherical_representation<degree> point2(45, 45, 3);
+
+    auto result = cross(point1, point2);
+
+    BOOST_CHECK_CLOSE(result.get_x().value(), -180.0, 0.001);
+    BOOST_CHECK_CLOSE(result.get_y().value(), 11.988, 0.001);
+    BOOST_CHECK_CLOSE(result.get_z().value(), -1485.0, 0.001);
+}
```
Comment (Ruchika2696) [inline test/coordinate/cartesian_representation.cpp:223]: add test for unit check

File: test/coordinate/cartesian_representation.cpp [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+{
+    auto point1 = make_cartesian_representation
+        (3.0 * meters, 5.0 * si::kilo *meters, 4.0 * si::mega * meters);
+    auto point2 = make_cartesian_representation
+        (3.0 * si::milli * meters, 5.0 * si::centi * meters, 4.0 * meters);
+
+    auto result = dot(point1, point2);
+
+    BOOST_CHECK_CLOSE(result.value(), 16000250009.0, 0.001);
+}
```
Comment (Ruchika2696) [inline test/coordinate/cartesian_representation.cpp:239]: add test for unit check

File: test/coordinate/cartesian_representation.cpp [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+
+BOOST_AUTO_TEST_CASE(magnitude)
+{
+    auto point1 = make_cartesian_representation
+        (25.0 * meter, 3600.0 * si::centi*meter, 90.0 * meter);
+
+    auto result = boost::astronomy::coordinate::magnitude(point1);
+
+    BOOST_CHECK_CLOSE(result.value(), 100.1049449328054, 0.001);
+
```
Comment (Ruchika2696) [inline test/coordinate/cartesian_representation.cpp:265]: Add test for unit check

File: test/coordinate/cartesian_representation.cpp [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+BOOST_AUTO_TEST_CASE(sum)
```
Comment (Ruchika2696) [inline test/coordinate/cartesian_representation.cpp]: Two tests for addition?

**Review by Ruchika2696 — CHANGES_REQUESTED** [review]
There are not tests for default constructor currently please add them too.

in differential file use `si::velocity` wherever possible and reformat the file.

File: test/coordinate/cartesian_representation.cpp (hunk 7 above)
Comment (sarthak2007) [reply test/coordinate/cartesian_representation.cpp]: Actually I forgot that unit conversion is not allowed for constructors

**Comment by sarthak2007** [comment]
> There are not tests for default constructor currently please add them too.
> 
> in differential file use `si::velocity` wherever possible and reformat the file.

Are you referring to this: https://github.com/BoostGSoC19/astronomy/pull/74/files#diff-3863ba304b9d2dc51d0427eea4f84174R92

**Pull request merged**